[Messenger] multiple failure transports support

monteiro opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #34911 
| License       | MIT
| Doc PR        | symfony/symfony-docs#13489
<!--
https://github.com/symfony/symfony-docs/pull/13489 Replace this notice by a short README for your feature/bugfix. This will help people
understand your PR and can be used as a start for the documentation.

Additionally (see https://symfony.com/roadmap):
 - Always add tests and ensure they pass.
 - Never break backward compatibility (see https://symfony.com/bc).
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against branch master.
-->

## Strategy applied

- Pass a map of transports and failed transports to the `SendFailedMessageToFailureTransportListener`. This way we re-use the same listener.
- Local failed transport has more priority than a global failed transport defined. 

## Configuration example

```yaml
# config/packages/messenger.yaml
framework:
    # no need to set failed transport globally if I want a specific behaviour per transport.
    failure_transport: failed # all transports have this failed transport
    messenger:
        transports:
            failed: 'doctrine://default?queue_name=failed'
            failed_important: 'doctrine://default?queue_name=failed_important'
            async:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                failure_transport: failed # takes precedence over the global defined "failed_transport"
                retry_strategy:
                    max_retries: 3
                    delay: 1000
                    multiplier: 2
            async_no_failure_transport: # it will use the global defined transport if no one is defined.
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                retry_strategy:
                    max_retries: 3
                    delay: 1000
                    multiplier: 2
            async_send_specific_failure_queue:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                failed_transport: failed_important # takes precedence over the global defined "failed_transport"
                retry_strategy:
                    max_retries: 3
                    delay: 1000
                    multiplier: 2
```

You can test this feature easily on a [demo project](https://github.com/monteiro/messenger-multiple-failed-transports-pr34979). Just follow the [README](https://github.com/monteiro/messenger-multiple-failed-transports-pr34979).

**More information on issue #34911** 

## What needs to be done so this can be merged:
- [x] validate strategy 
- [ ] update src/**/CHANGELOG.md files
- [x] update tests to cover all cases
- [x] create doc PR

monteiro committed:
[Messenger] Multiple failure transports support

monteiro commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1845,8 +1847,12 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
             $container->getDefinition('messenger.failure.send_failed_message_to_failure_transport_listener')
-                ->replaceArgument(0, $senderReferences[$config['failure_transport']]);
+                ->replaceArgument(0, $failedTransports);
             $container->getDefinition('console.command.messenger_failed_messages_retry')
```
what can we do with the commands for the retries if we have now multiple failed transports.
Should I create one command per transport with a failed queue defined?

sroze commented on src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:
```diff
@@ -42,6 +41,10 @@ public function onMessageFailed(WorkerMessageFailedEvent $event)
         }
 
+        if (!isset($this->failureSenders[$event->getReceiverName()])) {
+            return;
```
You should throw an error here. Because you `array_fill` in the wiring, this should be all good 👍

monteiro replied:
I changed the logic now. Maybe throwing an error doesn't make sense now.

sroze commented on src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:
```diff
@@ -27,12 +26,12 @@
     private $logger;
 
-    public function __construct(SenderInterface $failureSender, LoggerInterface $logger = null)
+    public function __construct(array $failureSenders, LoggerInterface $logger = null)
```
This is a BC-break. Can you make sure to also allow `SenderInterface`, which would be used by all receivers?

monteiro replied:
What happens if there is no global failure transport, what should be the value of the SenderInterface?

`SendFailedMessageToFailureTransportListener.php` is instantiated only once. That means that this class is used by all transports to send messages to their failure transport (that can be a global or a local one defined in each transport).

Should I create one `SendFailedMessageToFailureTransportListener` for each failure transport? I would need an extra argument that tells me the transport name, so I can discard messages of different failed transports.

monteiro replied:
My proposal, which I'm going to do next:

The constructor of `SendFailedMessageToFailureTransportListener`:
`public function __construct(?SenderInterface $globalfailureSender, LoggerInterface $logger = null, ServiceLocator $failureSenders)`

Will guarantee that BC is not broken.

sroze commented on src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:
```diff
@@ -27,12 +26,12 @@
     private $logger;
 
-    public function __construct(SenderInterface $failureSender, LoggerInterface $logger = null)
+    public function __construct(array $failureSenders, LoggerInterface $logger = null)
```
Instead of being an `array`, can you use a service locator instead?

monteiro replied:
`$failureSenders` is an array with the transport name and its failure transport.
The service locator doesn't hold this logic.

monteiro replied:
I understand now the service locator concept 😄thanks!

monteiro commented:
@sroze thanks a lot for the review. 

Maybe the best way to remove BC, is actually to do a separated listener, but only when multiple ones are defined. What do you think?

monteiro commented:
@sroze Already addressed all the comments to not break BC.

I have some questions as well regarding the failed commands without breaking BC:
- console.command.messenger_failed_messages_retry
- console.command.messenger_failed_messages_show
- console.command.messenger_failed_messages_remove

What should we do about them. Probably the best idea is to create more commands with different failed _messenger transports_ since now it's only one.

monteiro commented:
The strategy I'm using for failed commands is:
- Pass a failed transport name to the command that is a required argument.
- Passing a service locator of all failed transports to know which failed transport to consume from.

What's missing:
- Unit tests
- More testing (show and retry working, missing remove command).
- Rebase with latest master

monteiro commented:
Can someone help me with the review? 

I was very careful in order to not break any command or any class by extending some classes with more constructor parameters or adding more options to commands (the failed commands).

giovannialbero1992 commented:
What is missing in this PR? Only Changelog updates?

monteiro commented:
@giovannialbero1992 and rebasing now that I noticed there are some conflicting files. 
Maybe @weaverryan can help me unblock and give me some hints on this one.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/console.xml:
```diff
@@ -119,20 +119,23 @@
             <argument type="service" id="messenger.routable_message_bus" />
             <argument type="service" id="event_dispatcher" />
             <argument type="service" id="logger" />
+            <argument id="messenger.failure_transports_by_name.locator" /> <!-- Failed transports service locator -->
```
This should be moved to the new PHP config file.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml:
```diff
@@ -83,6 +83,20 @@
 
         <service id="messenger.transport.sqs.factory" class="Symfony\Component\Messenger\Bridge\AmazonSqs\Transport\AmazonSqsTransportFactory"/>
 
+        <!-- failure transports locator per transport -->
```
This should be moved to the new PHP config file.

fabpot reviewed (changes requested):
Quick review (I've not had a look at the feature itself).

weaverryan commented on src/Symfony/Component/Messenger/Command/AbstractFailedMessagesCommand.php:
```diff
@@ -30,18 +31,27 @@ abstract class AbstractFailedMessagesCommand extends Command
+    private $failureTransports;
 
-    public function __construct(string $receiverName, ReceiverInterface $receiver)
+    public function __construct(?string $receiverName, ?ReceiverInterface $receiver, ServiceLocator $failureTransports)
```
The purpose of `$receiverName` now is to be the `$globalFailureReceiverName`. I think we should rename the variable for clarity. The idea would be that, *if* this is set AND no `--failed-transport=` option is passed, then we will use this automatically. If this is NOT set, then we could ask the user interactively which failure receiver they want to use.

For the 2nd and 3rd arguments, I think they should be "merged". What I mean is, the "default failure receiver" (if there is one) will be inside of the `ServiceLocator $failureTransports` (if it's not already, it should be). This means that we only need 1 argument here, not 2.

We also need to worry about backwards-compatibility and triggering deprecation notices. So I would:

A) Only have a 2nd argument (not a 2nd and third)
B) Remove the type-hint from the argument (so that it allows both a `ReceiverInterface` or a `ServiceLocator`
C) Inside the constructor, check to see if the argument is `!instanceof ServiceLocator`. In that case, trigger a deprecation warning that says that they SHOULD now pass a locator of failure transports. Then write some code to make it work both ways - one simple way might be to create your own `ServiceLocator` object if they're using the deprecated code path - maybe something like:

```
$this->failureTransports = new ServiceLocator([$this->receiverName => function() { return $failureTransports; }
```

I'm not sure if that code to instantiate the ServiceLocator is 100% correct. But the idea is to create a mini-container with the ONE receiver to make the deprecated code path work. Then the rest of the code can work the same.

monteiro replied:
Already talked with @weaverryan about the solution

weaverryan commented on src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php:
```diff
@@ -36,6 +36,7 @@ protected function configure(): void
             ->setDefinition([
                 new InputArgument('id', InputArgument::OPTIONAL, 'Specific message id to show'),
                 new InputOption('max', null, InputOption::VALUE_REQUIRED, 'Maximum number of messages to list', 50),
+                new InputOption('failed-transport', null, InputOption::VALUE_OPTIONAL, 'Use a specific failed transport'),
```
All of these should probably be `failure-transport`, right? That's what we call it in the config

weaverryan commented on src/Symfony/Component/Messenger/Command/AbstractFailedMessagesCommand.php:
```diff
@@ -30,18 +31,27 @@ abstract class AbstractFailedMessagesCommand extends Command
     }
 
-    protected function getReceiverName(): string
+    protected function getReceiverName(?string $name): string
```
Also, for the case of (B) (where we are using the global/default transport because no option was passed), we should print a message telling the user how to use different transports (if there are any):

> To use a different failure transport, re-run the command with --failure-transport=
> Available failure transports are: foo, baz, larry

weaverryan commented on src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:
```diff
@@ -27,12 +28,17 @@
+    private $globalfailureSender;
 
-    public function __construct(SenderInterface $failureSender, LoggerInterface $logger = null)
+    public function __construct(SenderInterface $globalfailureSender = null, LoggerInterface $logger = null, ServiceLocator $failureSenders = null)
```
I would use the same strategy here that I mentioned for the `AbstractFailedMesagesCommand`: do NOT add a new argument, but instead, remove the type-hint from this first argument, trigger a deprecation warning of the old value is passed, and write code to make the old code still work.

monteiro replied:
@weaverryan  I need to get the transport name from a SenderInterface. How can I get that? I need it, because in the event, we get the receiver name and get from the locator the sender to send to the failed transport.

weaverryan reviewed (commented):
We're on the right path!

fabpot commented:
@monteiro Do you think you will have time in the coming weeks to finish this PR based on the feedback? Or do you need some help?

monteiro commented:
@fabpot Absolutely. Already started again yesterday. This week I will already have the comments applied.

monteiro committed:
improved solution using one argument with 2 types in listener

monteiro committed:
fix

monteiro committed:
fix

monteiro committed:
cs fix

monteiro committed:
add another service locator for the listener

monteiro committed:
cs fix

monteiro committed:
cs fix

monteiro committed:
fix

monteiro committed:
fix retry command

monteiro commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml:
```diff
@@ -83,6 +83,20 @@
+            <tag name="container.service_locator" />
+            <argument type="collection" /> <!-- failure transports map by transport -->
+            <argument /> <!-- failure transports locator -->
+        </service>
```
As I have discussed with you, we need two locators. 
One for the listener that needs to know the failed transport from the queue, and another one to retry a specific message on a specific failed transport.

monteiro committed:
fix

monteiro commented:
@fabpot @weaverryan This PR is ready to be reviewed. 

I created a symfony project [project](https://github.com/symfony/symfony/pull/34979), so anyone can test, even if you don't have the symfony repository downloaded.

monteiro commented on src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:
```diff
@@ -27,12 +28,14 @@
     private $logger;
 
-    public function __construct(SenderInterface $failureSender, LoggerInterface $logger = null)
+    public function __construct(?SenderInterface $globalFailureSender, LoggerInterface $logger = null, ?ServiceLocator $failureSenders = null)
```
I will need the `globalFailureSender` here, because `failureSenders` is map where key is the transport name. 
When setting a global `failure_transport` is not associated to a transport name.

I need it, because in https://github.com/symfony/symfony/pull/34979/files#diff-2eca49162e4b0a72c3825b50f160d944R72 we will ask the failure sender from the transport the event was consumed.

weaverryan replied:
Hmm. But could we make `$failureSenders` a FULL array of "receivers' => $failureSenderService.

Take this pseudo-config:

```yml
failure_transport: failure1
transports:
    async1: null
    async2:
       failure_transport: failure2
    failure1: null
    failure2: null
```

In this case, $failureSenders would be a map containing this:

```
[
    async1 => <failure1 sender service>
    async2 => <failure2 sender service>
    failure1 => <failure1 sender service>
    failure2 => <failure2 sender service>
]
```

Then we would have all the info we need in the map, correct? This "occurred" to me by looking *only* at this class (and not thinking about the rest of the code): in this class, it *feels* like we should be able to pass this object all the info it needs with one, full service locator of failure senders.

monteiro replied:
How do you which one is the global one? Because your map just says that one specific transport is assigned to one failure sender service. 

Specify a specific key for global? 

[
    "messenger_global" => <failure1 sender service>
    async1 => <failure1 sender service>
    async2 => <failure2 sender service>
    failure1 => <failure1 sender service>
    failure2 => <failure2 sender service>
]

weaverryan commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1863,24 +1863,59 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
             $container->getDefinition('console.command.messenger_failed_messages_show')
-                ->replaceArgument(0, $config['failure_transport']);
+                ->replaceArgument(0, $globalFailureReceiver)
+                ->replaceArgument(1, $container->getDefinition($failureTransportsServiceLocatorId));
```
these should be references, right? We don't normally set a Definition directly on an argument. And should it use `$failureTransportsServiceLocatorId` or `$failureTransportsServiceLocator` - I'm getting lost (this stuff confuses me) in which is which and what should be used.

weaverryan commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1863,24 +1863,59 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
-                ->replaceArgument(0, $config['failure_transport']);
+                ->replaceArgument(0, $globalFailureReceiver)
+                ->replaceArgument(1, $senderReferences[$config['failure_transport']] ?? null)
+                ->replaceArgument(5, $container->getDefinition($failureTransportsServiceLocator));
```
Same problem here (ish) as below - we should pass a Reference (which `$failureTransportsServiceLocator` is), not a Definition.

weaverryan commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1863,24 +1863,59 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
             $container->removeDefinition('console.command.messenger_failed_messages_retry');
             $container->removeDefinition('console.command.messenger_failed_messages_show');
             $container->removeDefinition('console.command.messenger_failed_messages_remove');
+            $container->removeDefinition('messenger.failure.send_failed_message_to_failure_transport_listener');
```
we could revert this change to lessen the diff

weaverryan commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1863,24 +1863,59 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
+            if ($transport['failure_transport']) {
+                if (!isset($config['transports'][$transport['failure_transport']])) {
+                    throw new LogicException(sprintf('Invalid Messenger configuration: the failure transport "%s" is not a valid transport or service id.', $transport['failure_transport']));
+                }
```
Should we be checking for `!isset($senderReferences[$transport['failure_transport']])` here instead? That would be consistent with the above code. It's more confusing, but there is the edge case that the failure_transport is not a registered transport, but just a "sender service id"

weaverryan commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php:
```diff
@@ -130,6 +131,18 @@
+                abstract_arg('failed transports map by transport name'),
+            ])
+            ->tag('container.service_locator')
+
```
If I'm correct in `FrameworkExtension`, these won't be needed.

weaverryan commented on src/Symfony/Component/Messenger/Command/FailedMessagesRetryCommand.php:
```diff
@@ -39,14 +40,16 @@ class FailedMessagesRetryCommand extends AbstractFailedMessagesCommand
+    private $globalReceiverName;
 
-    public function __construct(string $receiverName, ReceiverInterface $receiver, MessageBusInterface $messageBus, EventDispatcherInterface $eventDispatcher, LoggerInterface $logger = null)
+    public function __construct(?string $globalReceiverName, ?ReceiverInterface $globalReceiver, MessageBusInterface $messageBus, EventDispatcherInterface $eventDispatcher, LoggerInterface $logger = null, ?ServiceLocator $failureTransports = null)
```
Could we eliminate the new `$failureTransports` argument and instead remove the type-hint from `?ReceiverInterface $globalReceiver`? The 2nd arg would then be able to receive a `ReceiverInterface` OR a `ServiceLocator` and we would check the type and act accordingly. Maybe we would just pass that straight to the parent constructor, since it already has the deprecation.

weaverryan commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -298,18 +298,6 @@ private function registerReceivers(ContainerBuilder $container, array $busIds)
-                $definition = $container->getDefinition($failedCommandId);
-                $definition->replaceArgument(1, $receiverMapping[$definition->getArgument(0)]);
-            }
-        }
```
Hmm, wait a second. This is subtle. But currently, in `FrameworkExtension`, you are populating arg 0 with `$senderReferences[$config['failure_transport']]`. But before, we were populating them with *receiver* references. Even for me, this makes my head spin a bit :P. You could technically have a failure transport "receiver" service but no "sender" service... though this is (at the very least) an edge case. Usually a sender & receiver are a "pair" because you've defined them as a transport. But this doesn't need to be the case.

The situation would be this: I create a `ReceiverInterface` service and tag it with `messenger.receiver` and `alias` `my_failure_receiver`. But, I do not actually register this as a transport. Then, at the command line, I use `messenger:failure:show --failure-transport=my_failure_receiver`. That is technically legal. And it's even possible (though unlikely) to set up this kind of situation if you're using AMQP (e.g. you actually set up a true `failure_transport` called `failure1`, but then with custom binding & routing rules, those messages ultimately end up in a queue connected with the custom `my_failure_receiver` receiver.

So.... that's why this code was originally here and I think it probably still needs to be. I don't think that's a problem, other than some of the new code in `FrameworkExtension` should actually live here.

monteiro committed:
fix SendFailedMessageToFailureTransportListenerTest

monteiro committed:
remove failureTransports last argument of FailedMessagesRetryCommand

monteiro committed:
php cs fix

monteiro committed:
wip

fabpot commented:
We've just moved away from `master` as the main branch to use `5.x` instead. Unfortunately, I cannot reopen the PR and change the target branch to `5.x`. Can you open a new PR referencing this one to not loose the discussion? Thank you for your understanding and for your help.

monteiro commented:
opened new PR https://github.com/symfony/symfony/pull/38468 with 5.x as base

luigi370 commented:
This is just merged into symfony 5 right? is there any change to make it work on symfony 4.4 without updating the framework? thanks!

stof commented:
updating the framework...
Old versions of Symfony don't get the new features of newer versions (that would defeat the purpose of versioning)
